ddl/integration_test.go

```diff
@@ -59,3 +59,24 @@ func TestDefaultValueIsBinaryString(t *testing.T) {
 	tk.MustGetErrMsg("create table t (a blob default 0xE4BDA0E5A5BD81);",
 		"[ddl:1101]BLOB/TEXT/JSON column 'a' can't have a default value")
 }
+
+// https://github.com/pingcap/tidb/issues/30740.
+func TestDefaultValueInEnum(t *testing.T) {
+	collate.SetCharsetFeatEnabledForTest(true)
+	defer collate.SetCharsetFeatEnabledForTest(false)
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test;")
+	// The value 0x91 should not cause panic.
+	tk.MustExec("create table t(a enum('a', 0x91) charset gbk);")
+	tk.MustExec("insert into t values (1), (2);")                 // Use 1-base index to locate the value.
+	tk.MustQuery("select a from t;").Check(testkit.Rows("a", "")) // 0x91 is truncate.
+	tk.MustExec("drop table t;")
+	tk.MustGetErrMsg("create table t(a set('a', 0x91, '') charset gbk);",
```

Comment:
It's not same as MySQL. Although MySQL behaves strangely.
```mysql
mysql> create table t(a set('a', 0x91, '') charset gbk);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (1), (2), (3);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select * from t;
+------+
| a    |
+------+
| a    |
|      |
| a,   |
+------+
3 rows in set (0.00 sec)
```

Reply:
I believe this is a bug in MySQL because the result of `show create table` should success:

```sql
mysql> create table t (a set('a', 0x91, '') charset gbk);
Query OK, 0 rows affected (0.05 sec)

mysql> show create table t;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` set('a','','') CHARACTER SET gbk COLLATE gbk_chinese_ci DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> drop table t;
Query OK, 0 rows affected (0.04 sec)

mysql> CREATE TABLE `t` (
    ->   `a` set('a','','') CHARACTER SET gbk COLLATE gbk_chinese_ci DEFAULT NULL
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
ERROR 1291 (HY000): Column 'a' has duplicated value '' in SET
```